common/sgx/quote.c

```diff
@@ -302,3 +376,311 @@ oe_result_t oe_verify_quote_internal(
+{
+    if (oe_datetime_compare(from, latest_from) > 0)
+    {
+        *latest_from = *from;
+    }
+
+    if (oe_datetime_compare(until, earliest_until) < 0)
+    {
+        *earliest_until = *until;
+    }
+}
+
+oe_result_t oe_verify_quote_internal_with_collaterals(
+    const uint8_t* quote,
+    size_t quote_size,
+    const uint8_t* collaterals,
+    size_t collaterals_size,
+    oe_datetime_t* validation_time)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    uint8_t* local_collaterals = NULL;
+    size_t local_collaterals_size = 0;
+    oe_datetime_t validity_from = {0};
+    oe_datetime_t validity_until = {0};
+    oe_datetime_t creation_time = {0};
+
+    bool no_collaterals = false;
+
+    if (quote == NULL)
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    if (collaterals == NULL)
+    {
+        result = oe_get_collaterals_internal(
+            quote,
+            quote_size,
+            (uint8_t**)&local_collaterals,
+            &local_collaterals_size);
+
+        if (result == OE_QUOTE_PROVIDER_CALL_ERROR)
+        {
+            // No qe_identity info returned from the quote provider, this could
+            // be because either get_qe_identity_info API was not supported or
+            // unexpected error. In both cases, check against hardcoded quoting
+            // enclave properties instead Assert that the qe report's MRSIGNER
+            // matches Intel's quoting. We will remove these hardcoded values
+            // once the libdcap_quoteprov.so was updated to support qe identity
+            // feature.
+            no_collaterals = true;
+        }
+        else
+        {
+            OE_CHECK_MSG(
+                result, "Failed to get collaterals. %s", oe_result_str(result));
+            collaterals = local_collaterals;
+            collaterals_size = local_collaterals_size;
+        }
+    }
+
+    OE_CHECK_MSG(
+        oe_verify_quote_internal(quote, quote_size, no_collaterals),
+        "Failed to verify remote quote.",
+        NULL);
+
+    if (!no_collaterals)
+    {
+        oe_collaterals_t* collaterals_body =
+            (oe_collaterals_t*)(collaterals + OE_COLLATERALS_HEADER_SIZE);
+
+        OE_CHECK_MSG(
+            oe_get_quote_validity_with_collaterals_internal(
+                quote,
+                quote_size,
+                collaterals,
+                collaterals_size,
+                &validity_from,
+                &validity_until),
+            "Failed to validate quote. %s",
+            oe_result_str(result));
+
+        // Verify quote/collaterals for the given time.  Use collateral creation
+        // time if one was not provided.
+        if (validation_time == NULL)
+        {
+            OE_CHECK_MSG(
+                oe_datetime_from_string(
+                    collaterals_body->creation_datetime,
+                    sizeof(collaterals_body->creation_datetime),
+                    &creation_time),
+                "Invalid creation time in collaterals.",
```

Comment:
It might be a bad idea to spew out the  contents of the creation_datetime in the error case.

Reply:
This function is converting the string repressentation of the datetime to an actual oe_datetime_t.  There is no printing of the creation_datetime.

Reply:
It might NOT be a bad idea to spew out the contents of the creation_datetime in the error case.

Reply:
Makes sense. Thanks.